Data formats set of guides (#682)
ClickHouse/clickhouse-docs: docs/en/integrations/data-ingestion/data-formats/binary.md

---
sidebar_label: Binary and native
sidebar_position: 5
slug: /en/integrations/data-formats/binary-native
---

# Using native and binary formats in ClickHouse

ClickHouse supports multiple binary formats, which result in better performance and space efficiency. Binary formats are also safe in character encoding since data is saved in a binary form.

> 

Reviewer: should we populate `some_data` here?

Author: Sounds reasonable, I added SQL/TSV to the assets folder and a sentence to reference it.

│ hits  │ UInt32 │              │                    │         │                  │                │
└───────┴────────┴──────────────┴────────────────────┴─────────┴──────────────────┴────────────────┘
```

In production, we use `FROM INFILE` to import data:

```sql
INSERT INTO sometable
FROM INFILE 'data.clickhouse'
FORMAT Native
```

### Native format compression

We can also enable compression while exporting data to Native format (as well as most other formats) using a `COMPRESSION` clause:

```sql
SELECT * FROM some_data
INTO OUTFILE 'data.clickhouse'
COMPRESSION 'lz4'
FORMAT Native
```

We've used LZ4 compression for export. We'll have to specify it while importing data:

```sql
INSERT INTO sometable
FROM INFILE 'data.clickhouse'
COMPRESSION 'lz4'
FORMAT Native
```

## Exporting to RowBinary

Another binary format supported is [RowBinary](/docs/en/interfaces/formats.md/#rowbinary), which allows importing and exporting data in binary-represented rows:

```sql
SELECT * FROM some_data
INTO OUTFILE 'data.binary' FORMAT RowBinary
```

This will generate [data.binary](assets/data.binary) file in a binary rows format.

> ### Exploring RowBinary files

Author: The only reason I see is the reuse in different systems supporting the same format, but actually - a very good question! Thoughts are welcome.

Automatic schema inference is not supported for this format, so to explore before loading, we have to define schema explicitly:

```sql
SELECT *
FROM file('data.binary', RowBinary, 'path String, month Date, hits UInt32')
LIMIT 5

┌─path───────────────────────────┬──────month─┬─hits─┐
│ Bangor_City_Forest             │ 2015-07-01 │   34 │
│ Alireza_Afzal                  │ 2017-02-01 │   24 │
│ Akhaura-Laksam-Chittagong_Line │ 2015-09-01 │   30 │
│ 1973_National_500              │ 2017-10-01 │   80 │
│ Attachment                     │ 2017-09-01 │ 1356 │
└────────────────────────────────┴────────────┴──────┘
```

Consider using [RowBinaryWithNames](/docs/en/interfaces/formats.md/#rowbinarywithnames), which also adds a header row with a columns list. [RowBinaryWithNamesAndTypes](/docs/en/interfaces/formats.md/#rowbinarywithnamesandtypes) will also add an additional header row with column types.

### Importing from RowBinary files
To load data from a RowBinary file, we can use a `FROM INFILE` clause:

```sql
INSERT INTO sometable
FROM INFILE 'data.binary'
FORMAT RowBinary
```

## Importing single binary value using RawBLOB

> Suppose we want to read an entire binary file and save it into a field in a table. This is the case when [RawBLOB format](/docs/en/interfaces/formats.md/#rawblob) can be used. This format can be directly used with a single-column table only:

Author: fixed

```sql
CREATE TABLE images(data String) Engine = Memory
```

Let's save an image file to the `images` table:

```bash
cat image.jpg | clickhouse-client -q "INSERT INTO images FORMAT RawBLOB"
```

We can check the `data` field length which will be equal to the original file size:

```sql
SELECT length(data) FROM images

┌─length(data)─┐
│         6121 │
└──────────────┘
```

### Exporting RawBLOB data

This format can also be used to export data using an `INTO OUTFILE` clause:

```sql
SELECT * FROM images LIMIT 1
INTO OUTFILE 'out.jpg'
FORMAT RawBLOB
```

> Note that we had to `LIMIT 1` because exporting more than a single value will create a corrupted file.

Reviewer: @mrcrypster

## MessagePack

ClickHouse supports importing and exporting to [MessagePack](https://msgpack.org/) using [MsgPack](/docs/en/interfaces/formats.md/#msgpack). To export to MessagePack format:

```sql
SELECT *
FROM some_data
INTO OUTFILE 'data.msgpk'
FORMAT MsgPack
```

To import data from a [MessagePack file](assets/data.msgpk):

```sql
INSERT INTO sometable
FROM INFILE 'data.msgpk'
FORMAT MsgPack
```

## Protocol Buffers

To work with [Protocol Buffers](/docs/en/interfaces/formats.md/#protobuf) we first need to define a [schema file](assets/schema.proto):

```protobuf
syntax = "proto3";

message MessageType {
  string path = 1;
  date month = 2;
  uint32 hits = 3;
};
```

> Path to this schema file (`schema.proto` in our case) is set in a `format_schema` settings potion for [Protobuf](/docs/en/interfaces/formats.md/#protobuf) format:

Reviewer: @mrcrypster

```sql

> SELECT * FROM some_data

Reviewer: can the user reproduce this example? where is some_data?

Author: Used this table all over the docs, but didn't give a DDL/dump. Do you think I should mention it somewhere and let download data for it? It looks like table structure and data isn't important in such cases, since formatting clauses are focused on.

Reviewer: Maybe a intro page referencing where users can find the data and how to load it?

Reviewer: or just do what you've done in a lot of cases which is explicitly point out where they can find the data for each example. good with either.

INTO OUTFILE 'proto.bin'
FORMAT Protobuf
SETTINGS format_schema = 'schema:MessageType'
```

> We've saved data to [proto.bin](assets/proto.bin) file. ClickHouse also supports importing Protobuf data as well as nested messages. Consider using [ProtobufSingle](/docs/en/interfaces/formats.md/#protobufsingle) to work with a single Protocol Buffer message (length delimiters will be omitted in this case).

Reviewer: @mrcrypster

## Cap’n Proto

Another popular binary serialization format supported by ClickHouse is [Cap’n Proto](https://capnproto.org/). Similarly to `Protobuf` format, we have to define a schema file ([schema.capnp](assets/schema.capnp)) in our example:
```
@0xec8ff1a10aa10dbe;

struct PathStats {
  path @0 :Text;
  month @1 :UInt32;
  hits @2 :UInt32;
}
```

Now we can import and export using [CapnProto](/docs/en/interfaces/formats.md/#capnproto) format and this schema:

```sql
SELECT
    path,
    CAST(month, 'UInt32') AS month,
    hits

> FROM some_data

Reviewer: see above, where is this data?

```
@0xec8ff1a10aa10dbe;

struct PathStats {
  path @0 :Text;
  month @1 :UInt32;
  hits @2 :UInt32;
}
```

Now we can import and export using [CapnProto](/docs/en/interfaces/formats.md/#capnproto) format and this schema:

```sql
SELECT
    path,
    CAST(month, 'UInt32') AS month,
    hits
FROM some_data
INTO OUTFILE 'capnp.bin'
FORMAT CapnProto
SETTINGS format_schema = 'schema:PathStats'
```

Note that we had to cast the `Date` column as `UInt32` to [match corresponding types](/docs/en/interfaces/formats.md/#data_types-matching-capnproto).

## Other formats

ClickHouse introduces support for many formats, both text, and binary, to cover various scenarios and platforms. Explore more formats and ways to work with them in the following articles:

- [CSV and TSV formats](csv-tsv.md)
- [Parquet, Avro, Arrow and ORC](parquet-arrow-avro-orc.md)
- [JSON formats](json.md)
- [Regex and templates](templates-regex.md)
- **Native and binary formats**
- [SQL formats](sql.md)

> And also check [clickhouse-local](https://clickhouse.com/blog/extracting-converting-querying-local-files-with-sql-clickhouse-local) - a portable full-featured tool to work on local/remote files without the need for Clickhouse server.

Reviewer 2: ```suggestion
And also check [clickhouse-local](https://clickhouse.com/blog/extracting-converting-querying-local-files-with-sql-clickhouse-local) - a portable full-featured tool to work on local/remote files without starting ClickHouse server.
```

Reviewer 2: Capital H
